Modularity context

j-mracek opened the pull request:
Requires: https://github.com/rpm-software-management/ci-dnf-stack/pull/887

j-mracek committed:
Change usage of context and version in modular solver

Context will be used as key that represents a virtual repository in the
stream.

j-mracek committed:
Adjust modular solver to new context type

The new context allow us to install only the best modular candidate and
not a candidate for each version.

The change also allows to have broken the latest.

j-mracek committed:
Add ModulePackage.getStaticContext()

It is required for identification of context string meaning.

j-mracek committed:
Add getRequires() for ModulePackage

The original functionality was quite complicate using Dependency class.
The new function also formats a final string and can remove platform
dependencies.

j-mracek committed:
Make ModulePackageget::NameStream() as static function

It is required to allow to use it before ModulePackage is created.

j-mracek committed:
Add an alternative constructor for ModulePackage

It is required for MdDocument v2 to construct an alternative context
for modular solver.

j-mracek committed:
Add compatible layer for MdDocuments v2

It uses context according to V3 documents (same name, stream, and
requires) or it generates a new context from module requires.

j-mracek committed:
Fix modular queries with the new solver

j-mracek committed:
Improve formatting of error string for modules

With a new solver, modular solver could have an alternative context,
therefore it requires a different formatting procedure.

j-mracek commented:
@rh-atomic-bot try

j-mracek committed:
Call addVersion2Modules(); as late as possible

It will improve compatibility with v2 data and make mov3 convertor more
robust.

j-mracek committed:
Change mechanism of module conflicts

Originally each module package conflicted with all other streams. With
the new modular solver only one package is installed per stream
therefore module package could conflict with all other packages from
the same module.

The new approach is much more simple than the previous implementation.

j-mracek committed:
Fix load/update FailSafe

The change is required, because the structure of modular solvable was
changed.

j-mracek committed:
Decide how to handle context according to static_context value

The change is required because upstream libmodulemd was unable
to support V3 version without breakage of everything.

j-mracek committed:
Add additional fallback for module resolve

It allows to resolve dependencies when the latest has broken
dependencies but other modules from context not. It resolves the issue
when a release change dependencies. Now it will only shows a warning.

jrohel commented on libdnf/module/ModulePackage.hpp:
```diff
@@ -70,17 +70,32 @@ class ModulePackage // TODO inherit in future; : public Package
+     *
+     * @return bool
+     */
+    bool getStaticContext() const;
```
From my point of view, `isStaticContex()` is a more accurate name.

j-mracek commented on libdnf/module/ModulePackageContainer.cpp:
```diff
@@ -49,6 +49,42 @@ extern "C" {
 
+namespace {
+
+void goal2name_query(libdnf::Goal & goal, libdnf::Query & query)
```
Add comment what it does

j-mracek committed:
Report a new type of the module resolve error

The problem only in latest happens when dependency changed in module
name:stream:context but the formal dependency is already enabled.

j-mracek committed:
Remove unused code bump version

jrohel commented:
Good. Thanks for the hard work on modularity.
